src/openassetio-python/tests/bridge/python/test_converter.cpp

```diff
@@ -0,0 +1,268 @@
+#include <openassetio/hostApi/Manager.hpp>
+#include <openassetio/hostApi/ManagerFactory.hpp>
+#include <openassetio/hostApi/ManagerImplementationFactoryInterface.hpp>
+#include <openassetio/log/ConsoleLogger.hpp>
+#include <openassetio/log/LoggerInterface.hpp>
+#include <openassetio/log/SeverityFilter.hpp>
+#include <openassetio/managerApi/Host.hpp>
+#include <openassetio/managerApi/HostSession.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include <openassetio/managerApi/ManagerStateBase.hpp>
+
+namespace py = pybind11;
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace {
+constexpr const char* kTestTraitId = "TestTrait";
+}  // namespace
+
+SCENARIO("Mutations in one language are reflected in the other") {
+  // Cause the openassetio-python lib to be loaded so pybind can cast.
+  py::module_::import("openassetio");
+
+  GIVEN("A C++ object casted to a Python object") {
+    const TraitsDataPtr traitsData = TraitsData::make();
+    PyObject* pyTraitsData = openassetio::python::converter::castToPyObject(traitsData);
+    REQUIRE(pyTraitsData != nullptr);
+
+    WHEN("data is set via the C++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+
+  GIVEN("A Python object casted to a C++ object") {
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    const TraitsDataPtr traitsData =
+        openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+    WHEN("data is set via the C++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
```

Comment:
[question] Should this be up one level in the scope, as the incref happens at 69?

Reply:
I don't believe so, this is on the same level of scope as the incref, meaning it gets run after every test-leaf, as far as my understanding goes.

Reply:
Sorry misread the indents there, stands for cases below though I think.